File: src/1_connectivity_task/Python/1.3_weighted_fast_union.py
from sys import stdin

# Количество уникальных объектов (чисел)
N = 100

# Инициализация массивов
array = list(range(N))
sizes = [1] * N

# Основной цикл ввода данных (пар чисел) до достижения EOF (конца файла)
for line in stdin:
	p, q = [int(x) for x in line.split()]


	# Проверка, связаны ли объекты (опреация ПОИСК)
	i = p
	while i != array[i]:
		i = array[i]

	j = q
	while j != array[j]:
		j = array[j]

	if i == j:
		continue


	# Опреация ОБЪЕДИНЕНИЕ (корень МЕНЬШЕГО дерева ссылаем на корень БОЛЬШЕГО дерева, затем увеличиваем размер БОЛЬШЕГО дерева)
	if sizes[i] < sizes[j]:
		array[i] = j
		sizes[j] += sizes[i]
	else:
		array[j] = i
		sizes[i] += sizes[j]


	# Вывод пары чисел
	print(" {} {}".format(p, q))

File: src/1_connectivity_task/C++/1.3_weighted_fast_union.cpp
#include <iostream>

// Количество уникальных объектов (чисел)
const int N = 100;

int main(int, char**)
{
	// Объявление переменных
	unsigned int i, j, p, q, array[N], sizes[N];

	// Инициализация массивов
	for (i = 0; i < N; ++i) {
		array[i] = i;
		sizes[i] = 1;
	}

	// Основной цикл ввода данных (пар чисел) до достижения EOF (конца файла)
	while (std::cin >> p >> q) {

		// Проверка, связаны ли объекты (опреация ПОИСК)
		for (i = p; i != array[i]; i = array[i]);
		for (j = q; j != array[j]; j = array[j]);
		if (i == j) continue;


		// Опреация ОБЪЕДИНЕНИЕ (корень МЕНЬШЕГО дерева ссылаем на корень БОЛЬШЕГО дерева, затем увеличиваем размер БОЛЬШЕГО дерева)
		if (sizes[i] < sizes[j]) {
			array[i] = j;
			sizes[j] += sizes[i];
		} else {
			array[j] = i;
			sizes[i] += sizes[j];
		}


		// Вывод пары чисел
		std::cout << " " << p << " " << q << std::endl;
	}
}

